File: backend/routes/categories.py
"""Category CRUD routes."""

from flask import Blueprint, request, current_app
from flask_login import login_required, current_user
from sqlalchemy import func

from backend import db
from backend.api_response import ok_response, error_response
from backend.constants import UNCAT_NAME
from backend.models import Budget, Category, MemorizedTransaction, SavingsGoal, Transaction
from backend.lib.cache import cache_bust_dashboard_metrics
from backend.lib.categories import (
    find_category_by_name,
    get_uncategorized,
    list_categories_for_user,
)

bp = Blueprint("categories", __name__)


def _serialize_categories_with_counts(categories: list[Category], *, user_id: int) -> list[dict]:
    if not categories:
        return []

    category_ids = [int(c.id) for c in categories if c.id is not None]
    txn_counts = {
        int(category_id): int(count)
        for category_id, count in (
            db.session.query(
                Transaction.category_id,
                func.count(Transaction.id),
            )
            .filter(Transaction.user_id == user_id)
            .filter(Transaction.category_id.in_(category_ids))
            .group_by(Transaction.category_id)
            .all()
        )
    }

    items: list[dict] = []
    for c in categories:
        item = c.to_dict()
        item["transaction_count"] = txn_counts.get(int(c.id), 0)
        items.append(item)
    return items


def _dependent_counts(category_id: int, user_id: int) -> dict:
    """Return counts of all rows that reference this category."""
    txn_count = (
        Transaction.query
        .filter(Transaction.category_id == category_id, Transaction.user_id == user_id)
        .count()
    )
    budget_count = (
        Budget.query
        .filter(Budget.category_id == category_id, Budget.user_id == user_id)
        .count()
    )
    goal_count = (
        SavingsGoal.query
        .filter(SavingsGoal.linked_category_id == category_id, SavingsGoal.user_id == user_id)
        .count()
    )
    memorized_count = (
        MemorizedTransaction.query
        .filter(MemorizedTransaction.category_id == category_id, MemorizedTransaction.user_id == user_id)
        .count()
    )
    return {
        "transactions": txn_count,
        "budgets": budget_count,
        "goals": goal_count,
        "memorized": memorized_count,
    }


def _remap_category(source_id: int, target_id: int, user_id: int) -> dict:
    """Reassign all dependents from source category to target.

    Raises ValueError with a JSON-serialisable payload on budget conflict.
    Returns counts dict on success (does NOT commit).
    """
    conflicting_periods: list[str] = []
    source_budget_months = {
        row[0] for row in (
            Budget.query
            .with_entities(Budget.month)
            .filter(Budget.category_id == source_id, Budget.user_id == user_id)
            .all()
        )
    }
    if source_budget_months:
        target_budget_months = {
            row[0] for row in (
                Budget.query
                .with_entities(Budget.month)
                .filter(Budget.category_id == target_id, Budget.user_id == user_id)
                .all()
            )
        }
        conflicting_periods = sorted(source_budget_months & target_budget_months)

    if conflicting_periods:
        raise ValueError(
            {
                "error": "budget_conflict",
                "message": "Both categories have budgets for the same period(s). Resolve before reassigning.",
                "conflicting_periods": conflicting_periods,
            }
        )

    txn_count = (
        Transaction.query
        .filter(Transaction.user_id == user_id, Transaction.category_id == source_id)
        .update({Transaction.category_id: target_id}, synchronize_session=False)
    )
    budget_count = (
        Budget.query
        .filter(Budget.user_id == user_id, Budget.category_id == source_id)
        .update({Budget.category_id: target_id}, synchronize_session=False)
    )
    goal_count = (
        SavingsGoal.query
        .filter(SavingsGoal.user_id == user_id, SavingsGoal.linked_category_id == source_id)
        .update({SavingsGoal.linked_category_id: target_id}, synchronize_session=False)
    )
    memorized_count = (
        MemorizedTransaction.query
        .filter(MemorizedTransaction.user_id == user_id, MemorizedTransaction.category_id == source_id)
        .update({MemorizedTransaction.category_id: target_id}, synchronize_session=False)
    )
    return {
        "remapped_count": int(txn_count or 0),
        "budget_count": int(budget_count or 0),
        "goal_count": int(goal_count or 0),
        "memorized_count": int(memorized_count or 0),
    }


@bp.route("/api/categories", methods=["GET", "POST"])
@login_required
def api_categories():
    if request.method == "GET":
        cats = list_categories_for_user(current_user.id)
        items = _serialize_categories_with_counts(cats, user_id=int(current_user.id))
        return ok_response(data={"items": items}, legacy={"items": items})

    payload = request.get_json(silent=True) or {}
    name = (payload.get("name") or "").strip()
    if not name:
        return error_response("Name is required.", status=400, code="validation_error")
    if len(name) > 64:
        return error_response("Name too long (max 64 characters).", status=400, code="validation_error")
    is_income = bool(payload.get("is_income"))

    existing = find_category_by_name(name, current_user.id)
    if existing:
        item = existing.to_dict()
        return ok_response(data={"item": item}, legacy={"item": item}, status=200)

    cat = Category(name=name, user_id=current_user.id, is_income=is_income)
    db.session.add(cat)
    db.session.commit()
    item = cat.to_dict()
    return ok_response(data={"item": item}, legacy={"item": item}, status=201)


@bp.route("/api/categories/<int:cat_id>/delete", methods=["POST"])
@login_required
def api_delete_category(cat_id: int):
    """Delete a user-owned category.

    Behaviour:
      - System categories → 403.
      - No dependents → hard delete, 200.
      - Dependents present, no reassign_to → 409 with counts.
      - Dependents present, reassign_to provided → remap then delete, 200.
      - Budget conflict during remap → 409 with conflicting_periods.
    """
    cat = (
        Category.query
        .filter(Category.id == cat_id, Category.user_id == current_user.id)
        .first_or_404()
    )

    if cat.is_system:
        return error_response(
            f"'{cat.name}' is a system category and cannot be deleted.",
            status=403,
            code="system_category_protected",
        )

    payload = request.get_json(silent=True) or {}
    reassign_to = payload.get("reassign_to")

    counts = _dependent_counts(cat_id, current_user.id)
    has_dependents = any(counts.values())

    if has_dependents and reassign_to is None:
        return error_response(
            "This category has dependent rows. Provide 'reassign_to' to move them first.",
            status=409,
            code="has_dependents",
            extra={"dependent_counts": counts},
        )

    try:
        if has_dependents:
            try:
                target_id = int(reassign_to)
            except (TypeError, ValueError):
                return error_response("'reassign_to' must be a category id.", status=400, code="validation_error")

            if target_id == cat_id:
                return error_response("'reassign_to' must be a different category.", status=400, code="validation_error")

            target = (
                Category.query
                .filter(Category.id == target_id, Category.user_id == current_user.id)
                .first_or_404()
            )
            _ = target  # confirm it exists and belongs to the user

            try:
                _remap_category(cat_id, target_id, current_user.id)
            except ValueError as conflict:
                conflict_payload = conflict.args[0]
                return error_response(
                    conflict_payload["message"],
                    status=409,
                    code=conflict_payload["error"],
                    extra={"conflicting_periods": conflict_payload["conflicting_periods"]},
                )

        db.session.delete(cat)
        db.session.commit()
        cache_bust_dashboard_metrics(current_user.id)
        return ok_response(data={"deleted": True}, legacy={"deleted": True})

    except Exception:  # noqa: BLE001
        db.session.rollback()
        current_app.logger.exception("Category delete failed for cat_id=%s user_id=%s", cat_id, current_user.id)
        return error_response("Failed to delete category.", status=500, code="internal_error")


@bp.route("/api/categories/<int:source_id>/remap", methods=["POST"])
@login_required
def api_remap_category(source_id: int):
    """Remap all dependents from source category to target and optionally delete source."""
    payload = request.get_json(silent=True) or {}
    try:
        target_id = int(payload.get("target_id"))
    except (TypeError, ValueError):
        return error_response("target_id is required.", status=400, code="validation_error")

    if source_id == target_id:
        return error_response("source_id and target_id must be different.", status=400, code="validation_error")

    source = (
        Category.query
        .filter(Category.id == source_id, Category.user_id == current_user.id)
        .first_or_404()
    )
    target = (
        Category.query
        .filter(Category.id == target_id, Category.user_id == current_user.id)
        .first_or_404()
    )

    uncategorized = get_uncategorized(current_user.id)
    if int(source.id) == int(uncategorized.id) or source.name.lower() == UNCAT_NAME.lower():
        return error_response(
            "Cannot remap the Uncategorized category.",
            status=400,
            code="validation_error",
        )

    _ = target  # accessed only to verify ownership

    try:
        remap_counts = _remap_category(source_id, target_id, current_user.id)
    except ValueError as conflict:
        conflict_payload = conflict.args[0]
        return error_response(
            conflict_payload["message"],
            status=409,
            code=conflict_payload["error"],
            extra={"conflicting_periods": conflict_payload["conflicting_periods"]},
        )

    try:
        db.session.commit()
        cache_bust_dashboard_metrics(current_user.id)
        return ok_response(data=remap_counts, legacy=remap_counts)
    except Exception:  # noqa: BLE001
        db.session.rollback()
        current_app.logger.exception(
            "Category remap failed for source_id=%s target_id=%s user_id=%s",
            source_id, target_id, current_user.id,
        )
        return error_response("Failed to remap category.", status=500, code="internal_error")

File: apps/api/src/routes/categories.ts
// Response envelope contract (matches Flask api_response.py):
//   Success: { ok: true,  data: <resource>,  error: null, meta: {} }
//   Error:   { ok: false, data: null,         error: <string>, code: <string>, ...extras }
//
// Resource shapes for this module:
//   GET  /api/categories          → data: { items: CategoryItem[] }
//   POST /api/categories          → data: { item: CategoryItem }          status 201
//   DELETE /api/categories/:id    → data: { deleted: true }               ?reassign_to=<id>
//   POST /api/categories/:id/remap → data: RemapCounts
//
// Deviation from Flask POST /api/categories:
//   Flask silently returned 200 + existing category on duplicate name.
//   Here we return 409 "category_name_exists" with the existing item in the
//   payload so the frontend can recover without a follow-up GET.
//   POST means create-new; the 409 payload contains the existing item.

import { Hono } from "hono"
import { z } from "zod"
import { and, eq, inArray, sql } from "drizzle-orm"
import { getDb } from "../db/connection"
import { categories } from "../db/schema/categories"
import { transactions } from "../db/schema/transactions"
import { budgets } from "../db/schema/budgets"
import { savingsGoals } from "../db/schema/savings-goals"
import { memorizedTransactions } from "../db/schema/memorized-transactions"
import { requireAuth } from "../middleware/auth"
import { readRateLimit, writeRateLimit, heavyWriteRateLimit } from "../lib/rate-limit"

export const categoriesRouter = new Hono()

// ── Zod schemas ───────────────────────────────────────────────────────────────

const CreateBody = z.object({
  name: z
    .string()
    .min(1, "Name is required.")
    .max(64, "Name too long (max 64 characters).")
    .trim(),
  is_income: z.boolean().optional().default(false),
})

const RemapBody = z.object({
  target_id: z.number().int().positive("target_id must be a positive integer."),
})

// ── Types ─────────────────────────────────────────────────────────────────────

type CategoryItem = {
  id: number
  name: string
  is_income: boolean
  is_system: boolean
  transaction_count: number
}

type RemapCounts = {
  remapped_count: number
  budget_count: number
  goal_count: number
  memorized_count: number
}

function serializeCategory(
  row: typeof categories.$inferSelect,
  transactionCount = 0,
): CategoryItem {
  return {
    id: row.id,
    name: row.name,
    is_income: row.isIncome ?? false,
    is_system: row.isSystem,
    transaction_count: transactionCount,
  }
}

// ── GET /api/categories ───────────────────────────────────────────────────────

categoriesRouter.get("/", requireAuth, readRateLimit, async (c) => {
  const { userId } = c.get("session")
  const db = getDb()

  const cats = await db
    .select()
    .from(categories)
    .where(eq(categories.userId, userId))
    .orderBy(sql`LOWER(${categories.name})`, categories.id)

  let items: CategoryItem[]
  if (cats.length === 0) {
    items = []
  } else {
    const catIds = cats.map((cat) => cat.id)
    const countRows = await db
      .select({
        categoryId: transactions.categoryId,
        count: sql<number>`COUNT(${transactions.id})`,
      })
      .from(transactions)
      // transactions.categoryId is nullable; inArray skips NULLs at runtime
      .where(
        and(
          eq(transactions.userId, userId),
          inArray(transactions.categoryId as Parameters<typeof inArray>[0], catIds),
        ),
      )
      .groupBy(transactions.categoryId)

    const countMap = new Map(countRows.map((r) => [r.categoryId, Number(r.count)]))
    items = cats.map((cat) => serializeCategory(cat, countMap.get(cat.id) ?? 0))
  }

  return c.json({ ok: true, data: { items }, error: null, meta: {} })
})

// ── POST /api/categories ──────────────────────────────────────────────────────

categoriesRouter.post("/", requireAuth, writeRateLimit, async (c) => {
  const { userId } = c.get("session")

  const body = await c.req.json().catch(() => ({}))
  const parsed = CreateBody.safeParse(body)
  if (!parsed.success) {
    const msg = parsed.error.errors[0]?.message ?? "Validation error."
    return c.json({ ok: false, data: null, error: msg, code: "validation_error" }, 400)
  }
  const { name, is_income } = parsed.data

  const db = getDb()
  const [existing] = await db
    .select()
    .from(categories)
    .where(and(eq(categories.userId, userId), sql`LOWER(${categories.name}) = LOWER(${name})`))
    .limit(1)

  if (existing) {
    return c.json(
      {
        ok: false,
        data: null,
        error: `A category named '${existing.name}' already exists.`,
        code: "category_name_exists",
        existing_item: serializeCategory(existing),
      },
      409,
    )
  }

  const [{ id: newId }] = await db
    .insert(categories)
    .values({ userId, name, isIncome: is_income })
    .$returningId()

  const [created] = await db
    .select()
    .from(categories)
    .where(eq(categories.id, newId))
    .limit(1)

  return c.json(
    { ok: true, data: { item: serializeCategory(created) }, error: null, meta: {} },
    201,
  )
})

// ── DELETE /api/categories/:id ────────────────────────────────────────────────
// Dependent-row reassignment uses ?reassign_to=<id> query param — not a JSON
// body. DELETE-with-body is silently dropped by some proxies and CDNs, and
// query params are universally supported by HTTP clients and OpenAPI tooling.

categoriesRouter.delete("/:id", requireAuth, writeRateLimit, async (c) => {
  const id = Number(c.req.param("id"))
  if (!Number.isInteger(id) || id <= 0) {
    return c.json(
      { ok: false, data: null, error: "Invalid category id.", code: "validation_error" },
      400,
    )
  }
  const { userId } = c.get("session")
  const db = getDb()

  const [cat] = await db
    .select()
    .from(categories)
    .where(and(eq(categories.id, id), eq(categories.userId, userId)))
    .limit(1)

  if (!cat) {
    return c.json(
      { ok: false, data: null, error: "Category not found.", code: "not_found" },
      404,
    )
  }
  if (cat.isSystem) {
    return c.json(
      {
        ok: false,
        data: null,
        error: `'${cat.name}' is a system category and cannot be deleted.`,
        code: "system_category_protected",
      },
      403,
    )
  }

  const counts = await _getDependentCounts(db, id, userId)
  const hasDependents = Object.values(counts).some((n) => n > 0)

  const reassignParam = c.req.query("reassign_to")
  const reassignId = reassignParam !== undefined && reassignParam !== "" ? Number(reassignParam) : null

  if (hasDependents && reassignId === null) {
    return c.json(
      {
        ok: false,
        data: null,
        error: "This category has dependent rows. Provide 'reassign_to' to move them first.",
        code: "has_dependents",
        dependent_counts: counts,
      },
      409,
    )
  }

  if (reassignId !== null) {
    if (!Number.isInteger(reassignId) || reassignId <= 0) {
      return c.json(
        { ok: false, data: null, error: "'reassign_to' must be a valid category id.", code: "validation_error" },
        400,
      )
    }
    if (reassignId === id) {
      return c.json(
        { ok: false, data: null, error: "'reassign_to' must be a different category.", code: "validation_error" },
        400,
      )
    }
    const [target] = await db
      .select()
      .from(categories)
      .where(and(eq(categories.id, reassignId), eq(categories.userId, userId)))
      .limit(1)
    if (!target) {
      return c.json(
        { ok: false, data: null, error: "Reassignment target not found.", code: "not_found" },
        404,
      )
    }
  }

  try {
    await db.transaction(async (tx) => {
      if (hasDependents && reassignId !== null) {
        await _checkBudgetConflict(tx, id, reassignId, userId)
        await tx
          .update(transactions)
          .set({ categoryId: reassignId })
          .where(and(eq(transactions.userId, userId), eq(transactions.categoryId, id)))
        await tx
          .update(budgets)
          .set({ categoryId: reassignId })
          .where(and(eq(budgets.userId, userId), eq(budgets.categoryId, id)))
        await tx
          .update(savingsGoals)
          .set({ linkedCategoryId: reassignId })
          .where(and(eq(savingsGoals.userId, userId), eq(savingsGoals.linkedCategoryId, id)))
        await tx
          .update(memorizedTransactions)
          .set({ categoryId: reassignId })
          .where(and(eq(memorizedTransactions.userId, userId), eq(memorizedTransactions.categoryId, id)))
      }
      await tx.delete(categories).where(eq(categories.id, id))
    })
  } catch (err: unknown) {
    const e = err as { code?: string; message?: string; conflicting_periods?: string[] }
    if (e?.code === "budget_conflict") {
      return c.json(
        {
          ok: false,
          data: null,
          error: e.message ?? "Budget conflict.",
          code: "budget_conflict",
          conflicting_periods: e.conflicting_periods ?? [],
        },
        409,
      )
    }
    throw err
  }

  return c.json({ ok: true, data: { deleted: true }, error: null, meta: {} })
})

// ── POST /api/categories/:id/remap ────────────────────────────────────────────

categoriesRouter.post("/:id/remap", requireAuth, heavyWriteRateLimit, async (c) => {
  const sourceId = Number(c.req.param("id"))
  if (!Number.isInteger(sourceId) || sourceId <= 0) {
    return c.json(
      { ok: false, data: null, error: "Invalid category id.", code: "validation_error" },
      400,
    )
  }
  const { userId } = c.get("session")

  const body = await c.req.json().catch(() => ({}))
  const parsed = RemapBody.safeParse(body)
  if (!parsed.success) {
    return c.json(
      { ok: false, data: null, error: "target_id is required.", code: "validation_error" },
      400,
    )
  }
  const { target_id } = parsed.data

  if (sourceId === target_id) {
    return c.json(
      {
        ok: false,
        data: null,
        error: "source_id and target_id must be different.",
        code: "validation_error",
      },
      400,
    )
  }

  const db = getDb()
  const [[source], [target]] = await Promise.all([
    db
      .select()
      .from(categories)
      .where(and(eq(categories.id, sourceId), eq(categories.userId, userId)))
      .limit(1),
    db
      .select()
      .from(categories)
      .where(and(eq(categories.id, target_id), eq(categories.userId, userId)))
      .limit(1),
  ])

  if (!source) {
    return c.json(
      { ok: false, data: null, error: "Source category not found.", code: "not_found" },
      404,
    )
  }
  if (!target) {
    return c.json(
      { ok: false, data: null, error: "Target category not found.", code: "not_found" },
      404,
    )
  }
  if (source.isSystem) {
    return c.json(
      { ok: false, data: null, error: "Cannot remap a system category.", code: "validation_error" },
      400,
    )
  }

  let remapCounts!: RemapCounts
  try {
    await db.transaction(async (tx) => {
      await _checkBudgetConflict(tx, sourceId, target_id, userId)

      // Count rows before updating so the response reflects actual changes.
      const [[txnRow], [budRow], [goalRow], [memRow]] = await Promise.all([
        tx.select({ count: sql<number>`COUNT(*)` }).from(transactions).where(and(eq(transactions.userId, userId), eq(transactions.categoryId, sourceId))),
        tx.select({ count: sql<number>`COUNT(*)` }).from(budgets).where(and(eq(budgets.userId, userId), eq(budgets.categoryId, sourceId))),
        tx.select({ count: sql<number>`COUNT(*)` }).from(savingsGoals).where(and(eq(savingsGoals.userId, userId), eq(savingsGoals.linkedCategoryId, sourceId))),
        tx.select({ count: sql<number>`COUNT(*)` }).from(memorizedTransactions).where(and(eq(memorizedTransactions.userId, userId), eq(memorizedTransactions.categoryId, sourceId))),
      ])

      await tx
        .update(transactions)
        .set({ categoryId: target_id })
        .where(and(eq(transactions.userId, userId), eq(transactions.categoryId, sourceId)))
      await tx
        .update(budgets)
        .set({ categoryId: target_id })
        .where(and(eq(budgets.userId, userId), eq(budgets.categoryId, sourceId)))
      await tx
        .update(savingsGoals)
        .set({ linkedCategoryId: target_id })
        .where(and(eq(savingsGoals.userId, userId), eq(savingsGoals.linkedCategoryId, sourceId)))
      await tx
        .update(memorizedTransactions)
        .set({ categoryId: target_id })
        .where(and(eq(memorizedTransactions.userId, userId), eq(memorizedTransactions.categoryId, sourceId)))

      remapCounts = {
        remapped_count: Number(txnRow?.count ?? 0),
        budget_count: Number(budRow?.count ?? 0),
        goal_count: Number(goalRow?.count ?? 0),
        memorized_count: Number(memRow?.count ?? 0),
      }
    })
  } catch (err: unknown) {
    const e = err as { code?: string; message?: string; conflicting_periods?: string[] }
    if (e?.code === "budget_conflict") {
      return c.json(
        {
          ok: false,
          data: null,
          error: e.message ?? "Budget conflict.",
          code: "budget_conflict",
          conflicting_periods: e.conflicting_periods ?? [],
        },
        409,
      )
    }
    throw err
  }

  return c.json({ ok: true, data: remapCounts, error: null, meta: {} })
})

// ── Private helpers ───────────────────────────────────────────────────────────

type AnyDb = ReturnType<typeof getDb>

async function _getDependentCounts(
  db: AnyDb,
  categoryId: number,
  userId: number,
): Promise<{ transactions: number; budgets: number; goals: number; memorized: number }> {
  const [[txnRow], [budRow], [goalRow], [memRow]] = await Promise.all([
    db.select({ count: sql<number>`COUNT(*)` }).from(transactions).where(and(eq(transactions.userId, userId), eq(transactions.categoryId, categoryId))),
    db.select({ count: sql<number>`COUNT(*)` }).from(budgets).where(and(eq(budgets.userId, userId), eq(budgets.categoryId, categoryId))),
    db.select({ count: sql<number>`COUNT(*)` }).from(savingsGoals).where(and(eq(savingsGoals.userId, userId), eq(savingsGoals.linkedCategoryId, categoryId))),
    db.select({ count: sql<number>`COUNT(*)` }).from(memorizedTransactions).where(and(eq(memorizedTransactions.userId, userId), eq(memorizedTransactions.categoryId, categoryId))),
  ])
  return {
    transactions: Number(txnRow?.count ?? 0),
    budgets: Number(budRow?.count ?? 0),
    goals: Number(goalRow?.count ?? 0),
    memorized: Number(memRow?.count ?? 0),
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function _checkBudgetConflict(tx: any, sourceId: number, targetId: number, userId: number): Promise<void> {
  const [sourceBudgets, targetBudgets] = await Promise.all([
    tx.select({ month: budgets.month }).from(budgets).where(and(eq(budgets.userId, userId), eq(budgets.categoryId, sourceId))),
    tx.select({ month: budgets.month }).from(budgets).where(and(eq(budgets.userId, userId), eq(budgets.categoryId, targetId))),
  ])

  const sourceMonths = new Set((sourceBudgets as { month: string }[]).map((r) => r.month))
  const conflicting = (targetBudgets as { month: string }[])
    .map((r) => r.month)
    .filter((m) => sourceMonths.has(m))
    .sort()

  if (conflicting.length > 0) {
    throw Object.assign(
      new Error("Both categories have budgets for the same period(s). Resolve before reassigning."),
      { code: "budget_conflict", conflicting_periods: conflicting },
    )
  }
}
